graphistry/collapse.py

```diff
@@ -0,0 +1,532 @@
+                for e in get_edges_of_node(
+                    g, parent, directed=True
+                ).values:  # False just includes the child node
+                    collapse(
+                        g, e, child, attribute, column, seen
+                    )  # now child is the parent, and the edges are the start node
+        # else do nothing collapse-y to parent, move on to child
+        else:  # if (F, *)
+            #  do nothing to child, parent is child, and child is edge and recurse
+            for e in get_edges_of_node(g, child, directed=True).values:
+                if VERBOSE:
+                    print(
+                        f" -- Parent {parent} does not have property, looking at node <[ {e} from {child} ]>"
+                    )
+                collapse(
+                    g, e, child, attribute, column, seen
+                )  # now child is the parent, and the edges are the start node
+    return g
+
+
+def normalize_graph(
+    g, self_edges: bool = False, unwrap: bool = True, remove_collapse: bool = True
+):
+    """
+        Final step after collapse traversals are done, removes duplicates and moves COLLAPSE columns into respective
+        (node, src, dst) columns of node, edges dataframe from Graphistry instance g.
+
+    --------------------------------------------------------------------------------------------------------------------
+
+    :param g: graphistry instance
+    :param self_edges: bool, whether to keep duplicates from ndf, edf, default False
+    :param unwrap: bool, whether to unwrap node text with `~`, default True
+    :param remove_collapse: bool, whether to drop COLLAPSE columns from dataframe, default True
+    :returns final graphistry instance
+    """
+    # at the end of collapse, move anything untouched to new graph
+    ndf, edf, src, dst, node = unpack(g)
+
+    # move the new node names fromo COLLAPSE COL to the node column
+    ndf.loc[ndf[COLLAPSE_NODE] != DEFAULT_VAL, node] = ndf.loc[
+        ndf[COLLAPSE_NODE] != DEFAULT_VAL, COLLAPSE_NODE
+    ]
+    ndf = ndf.drop_duplicates()
+
+    edf.loc[edf[COLLAPSE_SRC] != DEFAULT_VAL, src] = edf.loc[
+        edf[COLLAPSE_SRC] != DEFAULT_VAL, COLLAPSE_SRC
+    ]
+    edf.loc[edf[COLLAPSE_DST] != DEFAULT_VAL, dst] = edf.loc[
+        edf[COLLAPSE_DST] != DEFAULT_VAL, COLLAPSE_DST
+    ]
+    if not self_edges:
+        edf = edf.drop_duplicates()
+
+    if unwrap:
+        ndf[node] = ndf[node].astype(str).apply(lambda x: unwrap_key(x))
+        edf[src] = edf[src].astype(str).apply(lambda x: unwrap_key(x))
+        edf[dst] = edf[dst].astype(str).apply(lambda x: unwrap_key(x))
+
+    if remove_collapse:
+        # remove collapse columns in ndf, edf
+        ndf = ndf.drop(columns=[COLLAPSE_NODE])
+        edf = edf.drop(columns=[COLLAPSE_SRC, COLLAPSE_DST])
+
+    # set the dataframes
+    g._nodes = ndf
+    g._edges = edf
+    return g
+
+
+def collapse_by(g, parent, start_node, attribute, column, seen):
```

Comment:
do you have a sense of how many ~seconds this is running on say:

- 100 node / 1K edges
- 1K nodes / 10K edges
- 10K nodes / 100K edges

Reply:
Get to this tomorrow in tests